### PR TITLE
Golangci-lint for all go-modules

### PR DESCRIPTION
**Pull request by mmorel-35**
In the actual workflow only the root module is linted.
This adds the scanning of all modules.
The following modules need to pass the linter to be enabled (removed from the exclusion list):
* examples/cockroachdb
* examples/spanner
* examples/toxiproxy
* modules/compose
* modules/pulsar
* modules/redis

I also noticed that the workflow was not benefiting from setup-go cache capability for two reasons. The first was that the checkout was done after the setup. The second reason was that the go.sum used as the cache dependency was not defined.

### CODE REVIEW

**Review by mdelapenya — DISMISSED** [review]
Great improvement, thanks!

LGMT

**Comment by mmorel-35** [comment]
Sorry for the dismiss of the review, I was fixing localstack.

It's ready to be merge on my part.

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -19,6 +19,34 @@ permissions:
 jobs:
   static-analysis:
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        working-directory:
+          - "."
+          - examples/bigtable
```
[the remark this answers is not in the document]
Comment (mmorel-35) [reply .github/workflows/golangci-lint.yml:27]: I agree. And also believe that every module/example shall be uncommented and fixed before adding it to the generation tool

**Review by mdelapenya — APPROVED** [review]
LGTM 🚀 

Let's include the YAML file to the code generation commands, but in a follow-up PR.

Thanks!

**Comment by mdelapenya** [comment]
@mmorel-35 Mmmm let me step back a little bit and share this thought.

I've just realised about the number of workers that we are going to consume at the moment we merge this, putting it in front of any PR: it will request 1 worker per module + example, 26 workers allocated at the moment. Even if they run super fast, it could drain the number of workers for the organisation. How would you see adding the golangci-lint as a step or reusable job, which is called for each module? We could include the golangci call into the `ci-test-go.yml` file as the first step. Wdyt?

**Review by mdelapenya — DISMISSED** [review]
Awesome, thanks!

image

File: .github/workflows/ci-test-go.yml [hunk 2]
```diff
@@ -57,6 +57,21 @@ jobs:
       - name: Check out code into the Go module directory
         uses: actions/checkout@v3
 
+      - name: golangci-lint
+        # TODO: Remove each example/module once it passes the golangci-lint
+        if: ${{ inputs.platform == 'ubuntu-latest' && !contains(fromJSON('["examples/cockroachdb","examples/spanner","examples/toxiproxy","modules/compose","modules/pulsar","modules/redis"]'), inputs.project-directory) }}
```
Comment (mmorel-35) [inline .github/workflows/ci-test-go.yml:62]: @mdelapenya , the condition for execution could be narrowed to enable less executions per module like on inputs.go-version or any other inputs.
What do you think ?

**Comment by mmorel-35** [comment]
Hi @mdelapenya ,
I have updated the description of the PR ti explain what it fully does. There is a flaky test unrelated to my modifications.

**Comment by mdelapenya** [comment]
> You might also need to update the PR rules

There is a repository `settings.yml` file including repo-as-code. You can update it yourself here in this PR: https://github.com/testcontainers/testcontainers-go/blob/71d58d8ec1ee541cb457ed40bdbfc287626aef24/.github/settings.yml#L186

I think removing the first string in the array would be enough

**Comment by mmorel-35** [comment]
Anything else you want to add here @mdelapenya ?

File: .golangci.yml [hunk 3]
```diff
@@ -1,16 +1,15 @@
 linters:
   enable:
     - gci
-    - gofmt
+    - gofumpt
```
Comment (mmorel-35) [inline .golangci.yml:4]: This is more strict linter than gofmt , it also ensure what has been talk [here](https://github.com/testcontainers/testcontainers-go/pull/1510#discussion_r1297170457)

File: .github/workflows/ci-test-go.yml [hunk 4]
```diff
@@ -48,14 +53,30 @@ jobs:
...
-            go-version: ${{ inputs.go-version }}
+            go-version: '${{ inputs.go-version }}'
+            cache-dependency-path: '${{ inputs.project-directory }}/go.sum'
         id: go
 
-      - name: Check out code into the Go module directory
-        uses: actions/checkout@v3
+      - name: golangci-lint
+        # TODO: Remove each example/module once it passes the golangci-lint
+        if: ${{ inputs.platform == 'ubuntu-latest' && inputs.go-version == '1.20.x' && !contains(fromJSON('["examples/cockroachdb", "examples/spanner", "examples/toxiproxy", "modules/compose", "modules/pulsar", "modules/redis"]'), inputs.project-directory) }}
```
Comment (mdelapenya) [inline .github/workflows/ci-test-go.yml:68]: What is the reason those modules need to be excluded? Are they hard to fix?

**Review by mdelapenya — APPROVED** [review]
LGTM, thanks for this huge work revisiting all files and updating them.

I wonder if we need a section in our docs regarding how to validate the format, e.g. running `golangci-lint run --verbose -c .golangci.yml`.

Do you think we can add it here in this PR, or in another one? No strong opinion on that, so whatever you prefer is fine.

In any case, thanks again for your work here, many times this kind of work is not visible at all, but super valuable, at least for me.

File: .github/workflows/ci-test-go.yml (hunk 4 above)
Comment (mmorel-35) [reply .github/workflows/ci-test-go.yml:68]: This PR has already several changes concerning the evolution of the workflow so I preferred not to include the fix of every linters for each module in there. That would be to much for me.

**Comment by mmorel-35** [comment]
> I wonder if we need a section in our docs regarding how to validate the format, e.g. running golangci-lint run --verbose -c .golangci.yml.


I think something can be added in the documentation as well as a tooling in the makefiles but in my opinion this is for another PR.
IMHO it's better to have many small PR doing one thing and doing it well than having one big.

File: .github/workflows/ci-test-go.yml (hunk 4 above)
Comment (mdelapenya) [reply .github/workflows/ci-test-go.yml:68]: Indeed, no questions about that 😊 just wanted to double check with you

**Comment by mdelapenya** [comment]
Merged, thanks!

image

**Pull request merged**